### PR TITLE
Clustering code cleanup

### PR DESCRIPTION
**Pull request by tumido**
I'm trying to make the clustering code readable. Let's simplify and optimize it. 

Related: ~https://github.com/numpy/numpy/issues/11999~ (Not any more relevant)

### CODE REVIEW

**Comment by tumido** [comment]
@durandom would you be interested in this type of PRs? I have also some questions about the current codebase - I'm placing them inline as this PR review.

File: clustering/__init__.py [hunk 1]
```diff
@@ -4,53 +4,25 @@
...
+    counts = np.count_nonzero(dataframe.loc[:, dataframe.columns.str.startswith('i_')], axis=1)
+    counts = counts.reshape(counts.shape[0],-1)
+    # Add the counts as a last column for each entry
+    return np.hstack((transformed, counts))
 
 def cluster(pd_rules):
-    # global incident_labels, m, n, pd_allin
-    np_rules = pd_rules.as_matrix()
-    labels = list(pd_rules.columns.values)
-    allin = np_rules[:, 3:]
```
Comment (tumido) [inline clustering/__init__.py]: Why are the first 3 columns extracted and not used?

The omitted columns are:
- `upload_time`
- `r_telemetry_rules_plugins_sap_sap_tmpfiles_report`
- `r_telemetry_rules_plugins_shift_ocp_service_denial_report`

I understand why is it so for the first one, not the others (since it continues with similar columns like `r_telemetry_rules_plugins_kernel_qlcnic_driver_crash_report`)

File: clustering/__init__.py [hunk 2]
```diff
@@ -4,53 +4,25 @@
...
+
+    # TODO: Fix me
     pd_allin = pd.DataFrame(data=np_allin[:, :-1], columns=allin_labels)
     # Kmeans clustering model pipeline
     min_max_scaler = preprocessing.StandardScaler()
     np_scaled = min_max_scaler.fit_transform(pd_allin)
     data = pd.DataFrame(np_scaled)
-    pca = PCA()
-    pca.fit(data)
     pca = PCA(n_components=3)
```
Comment (tumido) [inline clustering/__init__.py]: What's the rationale behind the number of components chosen to be 3? Also, why are we doing both PCA and K-Means? Isn't just one of these methods enough?

**Comment by tumido** [comment]
cc @bronaghs

File: clustering/__init__.py [hunk 3]
```diff
@@ -4,53 +4,25 @@
...
-            if allin[i][j] == True:
-                tmp.append(1)
-            elif allin[i][j] == False:
-                tmp.append(0)
-            elif allin[i][j] != 'None':
-                tmp.append(1)
-            else:
-                tmp.append(0)
-            if j in incident_labels and allin[i][j] == 1:
-                count += 1
```
Comment (tumido) [inline clustering/__init__.py]: Why are we counting the incident values? And why only if the input value is `1`? It doesn't make sense to have it restricted just to `1`s on the original input matrix. I'm not sure why to sum the values either? How does the clustering benefit from that since all the columns summed are already present? Was this verified to be needed?

Comment (tumido) [reply clustering/__init__.py]: Lol, I've just noticed, that the count is not used at all, since it's later removed on line [47](https://github.com/RedHatInsights/aicoe-insights-clustering/pull/9/files#diff-2a440676ce12f1a1b39194070ef96a56L47) via `np_allin[:, :-1]`. In that case I don't need the `counts` variable (in the new code) at all.

**Comment by tumido** [comment]
Also the numpy issue shouldn't matter anymore, since the missing data handling can be done in Pandas directly. I'll fix that in later iteration of this PR later today...

File: clustering/__init__.py [hunk 4]
```diff
@@ -4,53 +4,25 @@
...
-    incident_labels = find_incident(allin_labels)
-    m, n = allin.shape
-    np_allin = construct_np_allin(allin, m, n, incident_labels)
-    np_allin = np.array(np_allin)
+    np_allin = _create_incidents_matrix(pd_rules)
+
+    # TODO: Fix me
     pd_allin = pd.DataFrame(data=np_allin[:, :-1], columns=allin_labels)
     # Kmeans clustering model pipeline
     min_max_scaler = preprocessing.StandardScaler()
```
Comment (tumido) [inline clustering/__init__.py]: Why is the data scaled when it's just `True` or `False` values? And why is it scaled twice?

File: clustering/__init__.py [hunk 5]
```diff
@@ -4,53 +4,25 @@
...
-                tmp.append(1)
-            else:
-                tmp.append(0)
-            if j in incident_labels and allin[i][j] == 1:
-                count += 1
-        np_allin.append(tmp+[count])
-    return np_allin
-
+def _create_incidents_matrix(dataframe):
+    false_values = [False, 'False', None, 'None', 0]
```
Comment (Ladas) [inline clustering/__init__.py]: so  ```np.array([False, 'False', None, 'None', 0]).dtype``` is object, we should be safe here

maybe we want to set it explicitly though, to avoid the bug you see, where ```np.array([False, 'False']).dtype``` is >U4
so: ```np.array([False, 'False', None, 'None', 0], dtype=object)```
(also looking at how to set the exact type, since using object is supposed to be super slow)

Comment (tumido) [reply clustering/__init__.py]: No need for that... Since there's no longer the need for the numpy conversions I'm gonna use Pandas directly. And that doesn't suffers from the issue, so it's fine.

File: clustering/__init__.py (hunk 1 above)
Comment (durandom) [reply clustering/__init__.py]: thats probably an error. Or maybe the format is different on the anonymized data we have in the datahub?

```
❯ parquet-tools schema data/2018-07-19/rule_data/

message spark_schema {
  optional binary account (UTF8);
  optional binary system_id (UTF8);
  optional int96 upload_time;
  optional boolean c_telemetry_rules_plugins_non_kernel_rpm_database_problems_check_affected_version;
  optional boolean c_telemetry_rules_plugins_networking_vlan_interface_gateway_check_affected_version;
  optional boolean c_telemetry_rules_plugins_networking_neighbour_table_overflow_check_affected_version;
```

File: clustering/__init__.py (hunk 2 above)
Comment (durandom) [reply clustering/__init__.py]: AFAIR PCA is used to determine the number of clusters. So yes, PCA would not be needed here - or we determine it dynamically. 

@MichaelClifford do you know more?

File: clustering/__init__.py (hunk 3 above)
Comment (durandom) [reply clustering/__init__.py]: 🤣 yeah. As said, I just copy and pasted the code from the notebook.

File: clustering/__init__.py (hunk 1 above)
Comment (tumido) [reply clustering/__init__.py]: It seems so. The data I have downloaded doesn't have the `account` and `system_id` column.

**Comment by durandom** [comment]
@MichaelClifford could you do a review too?

File: clustering/__init__.py (hunk 1 above)
Comment (tumido) [reply clustering/__init__.py]: Correction: It does contain these columns, but for some reason they are placed as last ones.

Comment (tumido) [reply clustering/__init__.py]: Also, if we are transforming all the data to `True` and `False` values, it doesn't make any difference or bias if the account and system_id columns are kept there. All the values in these columns would be `True` no matter what... Removing columns is more expensive operation then just not paying attention to them.

**Comment by tumido** [comment]
@durandom, I've rewritten the rest:
- Since the PCA is not used much at this point, I've removed it. I can come up with a dynamic solution later. Let's focus on a static clustering for now.
- Also the `StandardScaler`s are removed since they make no sense over boolean values.
- I've skipped the column removal for `system_id` and `account` since the data are converted to boolean as well -> that way they do not bias the results and it would be much more expensive to remove them than just not paying attention to them.
- It gives me pretty good results and pretty fast. I need to do some measurements though.

@Ladas, here you go. This is the "Pandas only" approach I was talking about

File: clustering/__init__.py (hunk 2 above)
Comment (MichaelClifford) [reply clustering/__init__.py]: @durandom , @ tumido, 

It looks like PCA is only being used as a pre-processing step, reducing the dimensionality of the dataset from its original size down to just 3. It's not strictly required to perform the PCA prior to the K-means clustering, but its not unheard of.  If your results and performance are unaffected by removing it, then its probably fine to take out.

File: clustering/__init__.py [hunk 6]
```diff
@@ -1,67 +1,12 @@
...
-    pd_allin['cluster'] = kmeans[4].predict(data)
-    # pd_allin['principal_feature1'] = data[0]
-    # pd_allin['principal_feature2'] = data[1]
-    pd_allin['system_id'] = pd_rules['system_id']
-    return dict(zip(pd_allin['system_id'], pd_allin['cluster']))
+    Normalize data and than run KMeans on it (categorize into 5 clusters)
+    """
+    data = ~pd_rules.isin([False, 'False', None, 'None', 0])
+    clusters = KMeans(n_clusters=5).fit_predict(data)
+    return dict(zip(pd_rules['system_id'], clusters))
```
Comment (MichaelClifford) [inline clustering/__init__.py]: @tumido , @durandom This all looks fine to me from a DS perspective. 

The only things that I would note are that removing the "find_incident" function is changing the scope of the data being consumed here from just the "incident" records to all data fields (those that have the prefix "c_" and "r_"). Unless we are specifically concerned with "incident" clustering then I think this is fine, but thought it worth mentioning.

Second, if you do include all fields, they are not all boolean or None. Some are text. Your current implementation on line 10 converts any text value that is not None into True. These instances are fairly rare, and might not have too much of an impact on the overall clustering performance. But again, worth mentioning. They could be addressed with some categorical label encoding as a next step.

Comment (MichaelClifford) [reply clustering/__init__.py]: @tumido OK, my mistake. You are right, I thought ```construct_np_allin``` was constructing a dataset of only 'all incidents'. But I see now that's not the case, its still processing all fields. My bad. 

Yes, I agree. Let's experiment a bit, then we can refine from there. But as I said, current updates seem great. Far more streamlined.

File: clustering/__init__.py [hunk 7]
```diff
@@ -1,67 +1,12 @@
...
-    kmeans = [KMeans(n_clusters=i).fit(data) for i in n_cluster]
-    # The ultimate output: system with its associated cluster id -- pd_allin['cluster']
-    pd_allin['cluster'] = kmeans[4].predict(data)
-    # pd_allin['principal_feature1'] = data[0]
-    # pd_allin['principal_feature2'] = data[1]
-    pd_allin['system_id'] = pd_rules['system_id']
-    return dict(zip(pd_allin['system_id'], pd_allin['cluster']))
+    Normalize data and than run KMeans on it (categorize into 5 clusters)
+    """
+    data = ~pd_rules.isin([False, 'False', None, 'None', 0])
```
Comment (durandom) [inline clustering/__init__.py]: @tumido would you add your comment to this line? 

> Also, if we are transforming all the data to True and False values, it doesn't make any difference or bias if the account and system_id columns are kept there. All the values in these columns would be True no matter what... Removing columns is more expensive operation then just not paying attention to them.

File: clustering/__init__.py (hunk 6 above)
Comment (durandom) [reply clustering/__init__.py]: WOW. I thought the code was not good. But reducing it to 2 lines looks like magic :)

regarding the experimentation: I think we should start thinking about how we can make sure that changes like these dont impact the results/quality of the model.

@tumido we were experimenting with mlflow.org for that matter. have a look and let's discuss that next week
